File: articles/differential-equations-population-physiological-modeling/python/sir_epidemic.py
"""
SIR epidemic scenarios.

Run:
    python python/sir_epidemic.py
"""

from pathlib import Path

import pandas as pd

from differential_equations_core import SIRParameters, simulate_sir


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "sir_scenarios.csv"


def main() -> None:
    scenarios = pd.read_csv(DATA_PATH)
    rows = []

    for _, row in scenarios.iterrows():
        params = SIRParameters(
            row["beta"],
            row["gamma"],
            row["S0"],
            row["I0"],
            row["R0"],
            row["dt"],
            row["t_end"],
        )

        trajectory = simulate_sir(params)
        peak_idx = int(trajectory["infected"].idxmax())

        rows.append(
            {
                "scenario": row["scenario"],
                "R0_parameter_ratio": row["beta"] / row["gamma"],
                "peak_infected": trajectory["infected"].iloc[peak_idx],
                "time_to_peak": trajectory["time"].iloc[peak_idx],
                "final_recovered": trajectory["recovered"].iloc[-1],
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/differential-equations-population-physiological-modeling/r/sir_epidemic.R
# SIR epidemic model.

simulate_sir <- function(beta, gamma, S0, I0, R0, dt = 0.05, t_end = 120) {
  time <- seq(0, t_end, by = dt)
  S <- numeric(length(time))
  I <- numeric(length(time))
  R <- numeric(length(time))

  S[1] <- S0
  I[1] <- I0
  R[1] <- R0

  for (i in 2:length(time)) {
    dS <- -beta * S[i - 1] * I[i - 1]
    dI <- beta * S[i - 1] * I[i - 1] - gamma * I[i - 1]
    dR <- gamma * I[i - 1]

    S[i] <- max(S[i - 1] + dS * dt, 0)
    I[i] <- max(I[i - 1] + dI * dt, 0)
    R[i] <- max(R[i - 1] + dR * dt, 0)
  }

  data.frame(time = time, susceptible = S, infected = I, recovered = R)
}

trajectory <- simulate_sir(0.35, 0.10, 0.99, 0.01, 0)

peak_index <- which.max(trajectory$infected)

summary_df <- data.frame(
  peak_infected = trajectory$infected[peak_index],
  time_to_peak = trajectory$time[peak_index],
  final_recovered = tail(trajectory$recovered, 1)
)

print(round(summary_df, 5))
